add sections on key selections and cache behavior (#2)
thibmeu/draft-darling-key-directory-over-http: draft-darling-key-directory-over-http.md

* Should the draft define an IANA registry and require protocols to register
  their H

## Key Selection

The following is a deterministic algorithm for determining which Key a client
should use to fulfill their cryptographic needs. By using a deterministic
algorithm, Origins can more easily predict the effects of a Key rotation and
implement grace periods, soak times, etc., without modifications to a Key
Directory’s protocol representation.

Key Selection Algorithm:
1. Remove keys which cannot satisfy a request. E.g., their not-after fields are
   in the past, their not-before fields are in the future, or they don’t have
   the necessary cryptographic properties, etc.

> 2. If the not-before field exists, Sort the remaining keys by their not-before

Reviewer: ```suggestion
2. If a not-before field exists, sort the remaining keys by their not-before
```

   field in descending order.

> 3. Use the first key.

Reviewer: ```suggestion
3. Use the first key of the key directory, as it appears on the Key Directory format.
```

For protocols which define a not-before field, the above algorithm minimizes the

> chance that the used key has not expired between reading the directory and

Reviewer: clients will know if a key has expired or not because there is either an explicit not-after/expire header, or there is max-age which replaces it with `not-after=time_of_retrieval+max_age`

2. Clients on a scheduled rotation MAY be configured to distrust rotation outside
a fixed schedule. Protocols SHOULD define such policies.

## Cache behaviour

Caching the Key Directory lowers latency and reduces resource usage on the
Mirror and the Origin. An optimal caching strategy should minimize resource

> usage for both the Client and Origin and never have the client use an invalid

Reviewer: ```suggestion
usage for both the Client and Origin while preventing the client from using an invalid
```

key. 

These two requirements, minimizing resource usage and never using an invalid
key, are at odds with each other. In the event of an unplanned key rotation, a
client might use an invalid key. However, if a client fetched the keys for every

> request, it would waste time and network resources. 

Reviewer: while simple, I like that paragraph. it restates the goal

### not-before fields

Origins SHOULD, when possible in the Key Directory format, add a not-before
field or equivalent to each Key in the Directory. The not-before field allows
the directory to signal to a client when a Key is safe to use and allows for

> overlap in key rotations.

Reviewer: do we want to impose constraints on that not-before,, such as it being a unix timestamp in seconds?

Author: Yep, good point! Most expiration times are in seconds in any case. And providing ms precision for a resource fetched over a network isn't super useful.

### cache directives

Origins SHOULD respond with cache directives which control when the Key

> Directory should be revalidated. Origins SHOULD provide a max-age header which

Reviewer: ```suggestion
Directory should be revalidated. Origins SHOULD provide a Cache-Control header, or Expires header which
```

Reviewer: i bet there is an rfc about it

is slightly less than the grace period given for a key about to rotate. Clients
MUST respect the `max-age` cache directive and MAY respect other directives.

Mirrors MUST respect all provided cache directives. Mirrors SHOULD issue a
max-age value which matches the remaining time to live (TTL) value of the key
directory in the Mirror’s local cache.

> ### Client cache refresh

Reviewer: wouldn't this be covered by cache control `min-fresh` directive?

Reviewer 2: That would be covered if we prescribe the client send min-fresh, for a single cache system. For CDNs (or mirrors of mirrors) that use multiple caching layers im not sure this is covered.

Reviewer 2: as discussed, we should use s-maxage and have some verbiage
